### PR TITLE
Merge initial ram synchronization work

### PR DESCRIPTION
**Pull request by vivlim**

### CODE REVIEW

File: brains/andalite.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    def __init__(self, game, args=None):
+        Sapiens.__init__(self, game, args)
+        self.telepathy = Telepathy()
+
+    def Step(self):
+        self.sendMemory()
+        self.receiveBytes()
+        return Sapiens.Step(self)
+
+    def Event(self, evt):
```
Comment (lifning) [inline brains/andalite.py]: are Event, Path, Victory necessary here? (in fact, pretty sure Victory's gone from the base class)

File: brains/andalite.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+class Telepathy:
+    inboundMemoryWrites = queue.Queue()
+    outboundMemoryReads = queue.Queue()
+
+    def __init__(self):
+        thread = threading.Thread(target=self.threadMain, args=())
+        thread.daemon = True
+        thread.start()
+        #asyncio.get_event_loop().run_forever()
+    def threadMain(self):
```
Comment (lifning) [inline brains/andalite.py]: (PEP8) style nit: i think this method needs one more `\n` before it and one fewer `\n` after it

File: brains/andalite.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        loop = asyncio.new_event_loop()
+        #loop.run_in_executor(executor, self.handleSocket)
+        loop.create_task(self.handleSocket())
+        loop.run_forever()
+
+
+    async def handleSocket(self):
+        async with websockets.connect('ws://localhost:8765') as websocket:
```
Comment (lifning) [inline brains/andalite.py]: we should probably make the address be a parameter given to the brain (with this as the default)

File: brains/andalite.py (hunk 2 above)
Comment (lifning) [reply brains/andalite.py]: also it looks like all the new methods are lowerCamel while the existing stuff is UpperCamel and the PEP8 standard (as well as that of the libretro C API) is snake_case.  we should probably standardize on one style... can be decided later and done in a different commit tho

File: brains/andalite.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            done, pending = await asyncio.wait(
+                    [consumerTask, producerTask],
+                    return_when=asyncio.FIRST_COMPLETED)
+            for task in pending:
+                task.cancel()
+
+    async def messageConsumer(self, websocket):
+        while True:
+            message = await websocket.recv()
+            data = ast.literal_eval(message)
```
Comment (lifning) [inline brains/andalite.py:77]: this'll do for now, but we may want to think about a different wire format than hoping `eval(str(x))` remains an identity function for all the types.

File: brains/andalite.py (hunk 1 above)
Comment (vivlim) [reply brains/andalite.py]: nope i'll clean 'em up

File: brains/andalite.py (hunk 2 above)
Comment (vivlim) [reply brains/andalite.py]: agreed. let's snap to snake_case since that's more pythonic

Comment (vivlim) [reply brains/andalite.py]: created issue #4

File: brains/andalite.py (hunk 4 above)
Comment (vivlim) [reply brains/andalite.py:77]: agreed. this is also pretty inefficient tbh

Comment (vivlim) [reply brains/andalite.py:77]: created #5

File: brains/andalite.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        Sapiens.Event(self, evt)
+
+    def Path(self):
+        return Sapiens.Path(self)
+
+    def Victory(self):
+        return Sapiens.Victory(self)
+
+    def sendMemory(self):
+        self.sendNewBytes(0xcc24, 11) # menu data
```
Comment (lifning) [inline brains/andalite.py]: 0x11

Comment (vivlim) [reply brains/andalite.py]: carp 🐟

File: brains/andalite.py (hunk 3 above)
Comment (vivlim) [reply brains/andalite.py]: hard to do this with the way args are parsed... created #6 to track

File: brains/andalite.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    def receiveBytes(self):
+        try:
+            data = self.telepathy.inboundMemoryWrites.get_nowait()
+            start = data.get('offset')
+            lastData = self.lastSent.get(start)
+
+            if lastData == None or start != None and lastData.get('data') != data.get('data'):
+                print("writing data")
+                self.game.PokeMemoryRegion(data['offset'], data['data'])
+                self.lastSent['offset'] = data
```
Comment (lifning) [inline brains/andalite.py]: should this be self.lastSent[start]

**Review by lifning — APPROVED** [review]

**Pull request merged**